### PR TITLE
feature: Support for multiple endpoints by chainID

### LINKED ISSUE #73: feature: Support for multiple endpoints by chainID
With the current logic of multiple web3 providers support, if there is a web3 providers on the same chain declared twice, the last one will be the only stored.

A potential new feature that we might add in the future is to have a pool of web3 providers that automatically connect to the next if the current fail. You might take some ideas or code from the old ethservice queue we had when the Oracle was connecting to Ethereum (https://github.com/vocdoni/vocdoni-node/blob/release-azeno/service/ethevents.go#L109) and (https://github.com/vocdoni/vocdoni-node/blob/release-azeno/ethereum/handler/ethereumHandler.go#L90).

In any case, this is probably out of scope on this PR. However, we might start thinking about it; thus providers might be a map [chainID] -> Web3ProviderHandler

Where Web3ProviderHandle has a list of methods, such as:
```go
ChainID() int64
AddProvider(string)
NextProvider() ethClient
GetProvider() ethClient
String() string
```
The web3ProvidersHandler could make the Connect and check if the RPC endpoint actually works and is synced before returning it.

[TBD]

### PR DESCRIPTION
**Pull request by lucasmenendez**
- Includes an adaptation of https://github.com/vocdoni/vote-frame/blob/main/farcasterapi/web3/farcaster.go to handle multiple web3 endpoints by chain id.

Closes #73

### CODE REVIEW

File: scanner/providers/web3/networks.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+// chainID provided.
+func (nm *NetworksManager) EndpointsByChainID(chainID uint64) ([]*NetworkEndpoint, bool) {
+	nm.mtx.RLock()
+	defer nm.mtx.RUnlock()
+	endpoints, ok := nm.networks[chainID]
+	return endpoints, ok
+}
+
+// URIByChainID method returns the URI configured for the chainID provided.
+func (nm *NetworksManager) URIsByChainID(chainID uint64) ([]string, bool) {
```
Comment (p4u) [inline scanner/providers/web3/networks.go:165]: Rlock is missing

File: scanner/providers/web3/networks.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+		if data.ChainID == chainID {
+			return fmt.Sprintf("%s:%s", data.ShortName, hexAddress), true
+		}
+	}
+	return "", false
+}
+
+// String method returns a string representation of the *NetworksManager list.
+func (nm *NetworksManager) String() string {
+	shortNames := map[string]bool{}
```
Comment (p4u) [inline scanner/providers/web3/networks.go:202]: Rlock() missing

File: scanner/providers/web3/networks.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	for shortName := range shortNames {
+		shortNamesSlice = append(shortNamesSlice, shortName)
+	}
+	return fmt.Sprintf("%v", shortNamesSlice)
+}
+
+// CurrentBlockNumbers method returns a map of uint64-uint64, where the key is
+// the chainID and the value is the current block number of the network.
+func (nm *NetworksManager) CurrentBlockNumbers(ctx context.Context) (map[uint64]uint64, error) {
+	blockNumbers := make(map[uint64]uint64)
```
Comment (p4u) [inline scanner/providers/web3/networks.go:218]: Rlock() missing

**Review by p4u — CHANGES_REQUESTED** [review]
I'm not sure this PR adds the functionality discussed previously. Since it does not distribute the RPC requests by each call, but instead distributes clients.


Ideally, the networksManager would implement a type Client with the following interface (`networkmanager.Client(chainID)`):

```
// ContractCaller defines the methods needed to allow operating with a contract on a read
// only basis.
type ContractCaller interface {
	// CodeAt returns the code of the given account. This is needed to differentiate
	// between contract internal errors and the local chain being out of sync.
	CodeAt(ctx context.Context, contract common.Address, blockNumber *big.Int) ([]byte, error)

	// CallContract executes an Ethereum contract call with the specified data as the
	// input.
	CallContract(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
}
```

Then on each CodeAt or CallContract, behind the scenes, the networkManager chooses a different web3 endpoint from the list of working endpoints for such chainID.

erc721.NewERC721Contract() then receives the network manager client and the magic happens automatically. 

This is just an idea that tries to achieve the same functionality we have on the farcaster implementation. Feel free do discuss it.

PS: I think networksManager is a horrible name, web3pool or something like that would be more descriptive IMHO

File: scanner/providers/web3/web3_pool.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+	// get the endpoints for the chainID provided
+	endpoints, ok := nm.networks[chainID]
+	if !ok {
+		return nil, false
+	}
+	// check if there is an available endpoint and return it if found, after
+	// setting its available flag to false
+	for i, endpoint := range endpoints {
+		if endpoint.available {
+			nm.networks[chainID][i].available = false
```
[the remark this answers is not in the document]
Comment (lucasmenendez) [reply scanner/providers/web3/web3_pool.go:137]: Okey, i think now i got it.

**Review by p4u — APPROVED** [review]

**Pull request merged**